### PR TITLE
Allow the use of environment variables in .bowerrc

### PR DESCRIPTION
**Pull request by mithun**
Similar to npmrc

### CODE REVIEW

**Comment by mithun** [comment]
Added capabilities to _untildify_.

Helps with bower/bower#1784

**Comment by mithun** [comment]
@sheerun Can you please review this?

File: lib/util/expand.js [hunk 1]
```diff
@@ -19,8 +19,43 @@ function camelCase(config) {
...
+  var untildify = require('untildify');
+  f = untildify(f);
+
+  // replace any ${ENV} values with the appropriate environ.
+  var envExpr = /(\\*)\$\{([^}]+)\}/g;
+  return f.replace(envExpr, function (orig, esc, name) {
+    esc = esc.length && esc.length % 2;
+    if (esc) return orig;
+    if (undefined === process.env[name]) {
+      throw new Error('Failed to replace env in config: ' + orig);
```
Comment (sheerun) [inline lib/util/expand.js]: Can we be more descriptive here? Something like:

> 'Environment variable is not defined in ' + orig

File: lib/util/expand.js [hunk 2]
```diff
@@ -19,8 +19,43 @@ function camelCase(config) {
...
+
+function envReplace(config) {
+    var envReplaced = {};
+
+    object.forOwn(config, function (value, key) {
+        // Ignore null values
+        if (value == null) {
+            return;
+        }
+        envReplaced[key] = lang.isPlainObject(value) ? envReplace(value) : doEnvReplaceStr(value);
```
Comment (sheerun) [inline lib/util/expand.js]: Could we run doEnvReplaceStr on strings instead of non-objects?

File: lib/util/expand.js [hunk 3]
```diff
@@ -19,8 +19,43 @@ function camelCase(config) {
     return camelCased;
 }
 
+function doEnvReplaceStr (f) {
+  if (typeof f !== 'string' || !f) return f;
```
Comment (sheerun) [inline lib/util/expand.js]: We don't need this check if we tun this function only on strings.

**Comment by sheerun** [comment]
Hey @mithun. I've reviewed your code and it looks good :)

Could you fix my comments and squash the commits? I'll merge right away.

**Comment by sheerun** [comment]
btw. Would you like to join @bower/contributors ?

File: lib/util/expand.js [hunk 4]
```diff
@@ -19,8 +19,43 @@ function camelCase(config) {
     return camelCased;
 }
 
+function doEnvReplaceStr (f) {
```
Comment (zzarcon) [inline lib/util/expand.js]: Can you add some docs to this function? like what is this for and where we are using it? ^^

File: lib/util/expand.js (hunk 1 above)
Comment (zzarcon) [inline lib/util/expand.js]: +1 Also showing the variable name in the error will be useful

File: lib/util/expand.js [hunk 5]
```diff
@@ -19,8 +19,43 @@ function camelCase(config) {
...
+    if (esc) return orig;
+    if (undefined === process.env[name]) {
+      throw new Error('Failed to replace env in config: ' + orig);
+    }
+
+    return process.env[name];
+});
+}
+
+function envReplace(config) {
```
Comment (zzarcon) [inline lib/util/expand.js]: `envReplace` and `camelCase` are 90% the same, what do you think about creating a more generic function to convert the values? maybe something like

```convert(obj, 'env')```
```convert(obj, 'camel')```

Just a suggestion, maybe it's out of scope from this PR @sheerun

File: test/test.js [hunk 6]
```diff
@@ -212,7 +212,28 @@ describe('NPM Config on package.json', function () {
             assert.equal(process.env.HTTPS_PROXY, 'http://other-proxy.local');
         });
     });
+});
+
+describe('Allow ${ENV} variables in .bowerrc', function() {
+
+    it('sets values from process.env', function() {
+        process.env._BOWERRC_MY_PACKAGES = 'a';
+        process.env._BOWERRC_MY_TMP = '/tmp/b';
```
Comment (zzarcon) [inline test/test.js]: Can you assign the values `a` and `/tmp/b` to vars inside the `describe` body and latter just use it here and in the assert?

**Comment by zzarcon** [comment]
@mithun looking very good, just this small changes from my side and should be ok!

File: test/test.js (hunk 6 above)
Comment (sheerun) [inline test/test.js]: It's not necessary. They are local variables.

File: lib/util/expand.js (hunk 5 above)
Comment (sheerun) [inline lib/util/expand.js]: I think this is copy-paste from npm repository so at least we are sure it works.

Let it stay :)

File: lib/util/expand.js (hunk 1 above)
Comment (mithun) [inline lib/util/expand.js]: @sheerun  Sure

File: lib/util/expand.js (hunk 5 above)
Comment (mithun) [inline lib/util/expand.js]: @sheerun yes, this indeed is copy+pasted from npm ;-). I was looking for the exact same behavior.

**Comment by mithun** [comment]
@sheerun @zzarcon  Updated and squashed

**Comment by sheerun** [comment]
Looks good :) Thank you! I'll release it at the end of this week.

**Pull request merged**